src/interned.rs

```diff
@@ -396,33 +737,41 @@ where
         &self,
         db: &dyn Database,
         input: Id,
-        revision: Revision,
+        _revision: Revision,
         _cycle_heads: &mut CycleHeads,
     ) -> VerifyResult {
         let zalsa = db.zalsa();
+
+        // Record the current revision as active.
+        let current_revision = zalsa.current_revision();
+        self.revision_queue.record(current_revision);
+
         let value = zalsa.table().get::<Value<C>>(input);
-        if value.first_interned_at > revision {
-            // The slot was reused.
+
+        // SAFETY: `value.shard` is guaranteed to be in-bounds for `self.shards`.
+        let _shard = unsafe { self.shards.get_unchecked(value.shard as usize) }.lock();
+
+        // SAFETY: We hold the lock for the shard containing the value.
+        let value_shared = unsafe { &mut *value.shared.get() };
+
+        // The slot was reused.
+        if value_shared.id.generation() > input.generation() {
```

Comment:
Do we need this check, given that the `id` changes when the generation changes?

(which in turn means that no existing query can depend on the now "new" interned value)